presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -0,0 +1,203 @@
+        findAddressesRecursively(maxBitLength - 1, 0L, totalBitLength, addressRanges, fillInNulls(valueRanges), rangeType);
+        return addressRanges;
+    }
+
+    private List<ZValueRange> fillInNulls(List<ZValueRange> valueRanges)
+    {
+        List<ZValueRange> nonNullValues = new ArrayList<>(valueRanges.size());
+        for (int index = 0; index < valueRanges.size(); index++) {
+            ZValueRange range = valueRanges.get(index);
+
+            Integer minimumValue = range.getMinimumValue();
+            Integer maximumValue = range.getMaximumValue();
+
+            if (minimumValue == null) {
+                minimumValue = 0;
+            }
+            if (maximumValue == null) {
+                maximumValue = (1 << encodingBits.get(index)) - 1;
+            }
+
+            nonNullValues.add(new ZValueRange(Optional.of(minimumValue), Optional.of(maximumValue)));
+        }
+        return nonNullValues;
+    }
+
+    /**
+     * Recursive function to find all address ranges within given value ranges.
+     * <p/>
+     * Returns when the address range is either completely within or completely out of the value ranges.
+     * Recurses to a smaller Z-order curve if the address range only partially overlaps with the value ranges.
+     *
+     * @param level the level of recursion on the Z-curve
+     * @param startAddress the starting address of the current Z-curve level
+     * @param totalCurveBits the amount of bits in the current Z-curve level
+     */
+    private void findAddressesRecursively(int level, long startAddress, int totalCurveBits, List<ZAddressRange<Long>> addressRanges, List<ZValueRange> valueRanges, RangeType rangeType)
+    {
+        long endAddress = startAddress + ((long) 1 << totalCurveBits) - 1;
+
+        List<Integer> startValues = zOrder.decode(startAddress);
+        List<Integer> endValues = zOrder.decode(endAddress);
+
+        if (inRange(startValues, endValues, valueRanges)) {
+            addToAddressList(startAddress, endAddress, addressRanges, rangeType);
+            return;
+        }
+
+        if (outOfRange(startValues, endValues, valueRanges)) {
+            return;
+        }
+
+        totalCurveBits -= dimensions[level];
+        int subtrees = 1 << dimensions[level];
+        for (int i = 0; i < subtrees; i++) {
+            findAddressesRecursively(level - 1, startAddress, totalCurveBits, addressRanges, valueRanges, rangeType);
+            startAddress += ((long) 1 << totalCurveBits);
+        }
+    }
+
+    private void addToAddressList(long startAddress, long endAddress, List<ZAddressRange<Long>> addressRanges, RangeType rangeType)
+    {
+        if (rangeType == RangeType.INTEGER) {
+            checkArgument((startAddress <= Integer.MAX_VALUE) && (endAddress <= Integer.MAX_VALUE),
+                    format("The address range [%d, %d] contains addresses greater than integers.", startAddress, endAddress));
+        }
```

Comment:
Since rangeType is only used here, if we move this logic into "zOrderSearchCurveIntegers" function, do you think if it will be cleaner? Meanwhile, if we do that, we don't need rangeType variable at all.